src/util.c

```diff
@@ -456,3 +459,691 @@ jv jq_util_input_next_input(jq_util_input_state *state) {
+  size_t n = 0;
+
+  if ((charvec = calloc(jv_array_length(vec) + 1, sizeof(charvec[0])))) {
+    jv_array_foreach(vec, i, v) {
+      if (jv_get_kind(v) != JV_KIND_STRING) {
+        jv_free(v);
+        free(charvec);
+        *res = jv_invalid_with_msg(jv_string("argument vector must be an array of strings"));
+        return NULL;
+      }
+      charvec[i] = jv_string_value(v);
+      jv_free(v); // we still have a reference via vec
+      n++;
+    }
+    jv_free(vec);
+    charvec[n] = NULL;
+  }
+  return charvec;
+}
+
+jv jq_spawn_process(jq_state *jq, jv file, jv file_actions, jv attrs, jv argv, jv env) {
+  posix_spawn_file_actions_t pfacs;
+  posix_spawnattr_t pattrs;
+  struct pipes p;
+  const char **argvp = NULL;
+  const char **envp = NULL;
+  pid_t pid;
+  int ret;
+  jv res = jv_invalid();
+
+  p.pin[0] = p.pin[1] = p.pout[0] = p.pout[1] = p.perr[0] = p.perr[1] = -1;
+
+  if ((ret = posix_spawn_file_actions_init(&pfacs))) {
+    res = jv_invalid_with_msg(jv_string_fmt("Could not spawn process: %s", strerror(ret)));
+    goto out3;
+  }
+  if ((ret = posix_spawnattr_init(&pattrs))) {
+    res = jv_invalid_with_msg(jv_string_fmt("Could not spawn process: %s", strerror(ret)));
+    goto out2;
+  }
+
+  if (jv_get_kind(file) != JV_KIND_STRING) {
+    res = jv_invalid_with_msg(jv_string("spawn file must be a string"));
+    goto out;
+  }
+  if (jv_get_kind(argv) != JV_KIND_ARRAY) {
+    res = jv_invalid_with_msg(jv_string("argument vector must be an array of strings"));
+    goto out;
+  }
+  res = jq_io_policy_check(jq, JV_OBJECT(jv_string("kind"), jv_string("spawn"),
+                                         jv_string("file"), jv_copy(file),
+                                         jv_string("file_actions"), jv_copy(file_actions),
+                                         jv_string("attributes"), jv_copy(attrs),
+                                         jv_string("argv"), jv_copy(argv),
+                                         jv_string("env"), jv_copy(env)));
+  if (jv_get_kind(res) != JV_KIND_TRUE)
+    goto out;
+  if ((argvp = jv2charvec(jv_copy(argv), &res)) == NULL)
+    goto out;
+  if (jv_get_kind(env) != JV_KIND_NULL && jv_get_kind(env) != JV_KIND_ARRAY) {
+    res = jv_invalid_with_msg(jv_string("environment vector must be an array of strings"));
+    goto out;
+  }
+  if (jv_get_kind(env) == JV_KIND_ARRAY && (envp = jv2charvec(jv_copy(env), &res)) == NULL)
+    goto out;
+
+  if (jv_get_kind(file_actions) == JV_KIND_ARRAY) {
+    jv_array_foreach(file_actions, i, v) {
+      if (jv2pfac(jq, v, &p, &pfacs, &res))
+        goto out;
+    }
+  }
+
+  if (jv_get_kind(attrs) == JV_KIND_ARRAY) {
+    jv_array_foreach(attrs, i, v) {
+      if (jv2pattr(jv_copy(v), &pattrs, &res))
+        goto out;
+    }
+  }
+
+  if ((ret = posix_spawnp(&pid, jv_string_value(file), &pfacs, &pattrs,
+                          (char * const *)argvp,
+                          envp ? (char * const *)envp : environ))) {
```

Comment:
I have a compile error here with message `identifier "environ" is undefined`
adding a 
```
#ifndef WIN32
extern char **environ;
#endif
```
in front of the function fixes it.
Seems like in your case some defines are set which declare that variable within some headers?

Reply:
Yeah, on POSIX systems `<unistd.h>` defines it.

Reply:
So why didn't it work for me on macOS? I've performed a clean rebuild with `autoreconf; configure; make`

Reply:
When I get access to an OS X system I'll check.

Reply:
from [here](https://epics.anl.gov/tech-talk/2009/msg00299.php)
```
/* Starting in Mac OS X 10.5 (Leopard), shared libraries and
 * bundles don't have direct access to environ (man environ).
 *
 * Define environ as (*_NSGetEnviron())
 */
#include <crt_externs.h>
#define environ (*_NSGetEnviron())

```

it looks like for macs we should do 
```
 #if defined(__APPLE__)
 #  include <crt_externs.h>
 #  define environ (*_NSGetEnviron())
 #endif
```

Reply:
another [nice example](https://cpp.hotexamples.com/examples/-/-/_NSGetEnviron/cpp-_nsgetenviron-function-examples.html#0x29da595a9c894f0a37412b81d10fb3d69722e00643615575b48fd6645156550d-44,,69,)
![image](https://user-images.githubusercontent.com/3144935/72837372-0af86280-3c97-11ea-8901-4f68f4aa1a2b.png)

Reply:
I'm doing this in my local build:
```
#ifndef WIN32
#  ifdef __APPLE__
#    include <crt_externs.h>
#    define environ (*_NSGetEnviron())
#  else
     extern char ** environ;
#  endif
#endif
```